mathics/core/formatter.py

```diff
@@ -157,6 +167,215 @@ def add_conversion_fn(cls, module_fn_name=None) -> None:
+        else:
+            pexp = ""
+
+    # pad right with '0'.
+    if len(s) < exp + 1:
+        if evaluation is not None:
+            evaluation.message("NumberForm", "sigz")
+        # TODO NumberPadding?
+        s = s + "0" * (1 + exp - len(s))
+    # pad left with '0'.
+    if exp < 0:
+        s = "0" * (-exp) + s
+        exp = 0
+
+    # left and right of NumberPoint
+    left, right = s[: exp + 1], s[exp + 1 :]
+
+    def _round(number, ndigits):
+        """
+        python round() for integers but with correct rounding.
+        e.g. `_round(14225, -1)` is `14230` not `14220`.
+        """
+        assert isinstance(ndigits, int)
+        assert ndigits < 0
+        assert isinstance(number, int)
+        assert number >= 0
+        number += 5 * int(10 ** -(1 + ndigits))
+        number //= int(10**-ndigits)
+        return number
+
+    # pad with NumberPadding
+    if f is not None:
+        if len(right) < f:
+            # pad right
+            right = right + (f - len(right)) * options["NumberPadding"][1]
+        elif len(right) > f:
+            # round right
+            tmp = int(left + right)
+            tmp = _round(tmp, f - len(right))
+            tmp = str(tmp)
+            left, right = tmp[: exp + 1], tmp[exp + 1 :]
+
+    def split_string(s, start, step):
+        if start > 0:
+            yield s[:start]
+        for i in range(start, len(s), step):
+            yield s[i : i + step]
+
+    # insert NumberSeparator
+    digit_block = options["DigitBlock"]
+    if digit_block[0] != 0:
+        left = split_string(left, len(left) % digit_block[0], digit_block[0])
+        left = options["NumberSeparator"][0].join(left)
+    if digit_block[1] != 0:
+        right = split_string(right, 0, digit_block[1])
+        right = options["NumberSeparator"][1].join(right)
+
+    left_padding = 0
+    max_sign_len = max(len(options["NumberSigns"][0]), len(options["NumberSigns"][1]))
+    i = len(sign_prefix) + len(left) + len(right) - max_sign_len
+    if i < n:
+        left_padding = n - i
+    elif len(sign_prefix) < max_sign_len:
+        left_padding = max_sign_len - len(sign_prefix)
+    left_padding = left_padding * options["NumberPadding"][0]
+
+    # insert NumberPoint
+    if options["SignPadding"]:
+        prefix = sign_prefix + left_padding
+    else:
+        prefix = left_padding + sign_prefix
+
+    if is_int:
+        s = prefix + left
+    else:
+        s = prefix + left + options["NumberPoint"] + right
+
+    # base
+    base = "10"
+
+    # build number
+    method = options["NumberFormat"]
+    if options["_Form"] in ("System`InputForm", "System`FullForm"):
+        return method(
+            _BoxedString(s, number_as_text=True),
+            _BoxedString(base, number_as_text=True),
+            _BoxedString(pexp, number_as_text=True),
+            options,
+        )
+    else:
+        return method(_BoxedString(s), _BoxedString(base), _BoxedString(pexp), options)
+
+
 #
 #   Comment from Rocky:
```

Comment:
As far as I can tell, this comment is (now?) no longer near anything that seems relevant to what it is saying. For example, I don't see where we test for str-ness and convert to Symbol. I think I saw that somewhere the last time I saw this comment.

Reply:
Ah. Maybe it is supposed to go before `format_element()` ?

Reply:
Below, `_BoxedString()` doesn't belong here. In fact it is the only class where everything else is a function

The boxing class belongs in the box module like all of the others boxing routines. And I don't think the box module should be under `builtin`. 

Here is a suggested reorganization.  Under core, we have a `form` (both `convert` and `form` are intended to be imperative verbs not nouns as we did in summary text, as in: you (we, they, or I) convert, you (or we, they, or I) `form`. 

`parser` is a noun, but if things were done from scratch I would have suggested `parse` instead of `parser`.  If you think it worth doing, then please do. 

Under format we have "box" (also a verb), and  low-level or final format (as opposed to the top level form). Right now this is low-level format is not under core. But I think it should be moved into core.

The core module then has parse, evaluate (not currently a module though) and form modules. The form module has box and format modules.

 If you wanted to write a backend to handle some new `Form` you  would model it on the core form module. 

What do you think?

Reply:
Also thoughts about putting `form` up  at the same level as `core` and `builtin` (replacing `format` which gets pushed down a level into `form`)?

Reply:
@rocky, actually, all this PR was just about following the idea in https://github.com/Mathics3/mathics-core/discussions/317#discussioncomment-2865937 about following WR  to split and reorganize `inout.py`, and move as far as possible all the very low-level formatting code to `mathics.core`.  What I am trying to do now (in a private branch) is to try figuring out how all these parts are working now, and how they should really work. Once I get that, it is going to make more sense to write more extensive documentation. If you prefer, I can put this as a draft, or just remove the PR. However, as always that we try to split and reorganize code at this level, we arein the risk of making the effort useless because of merge conflicts.

Reply:
Thanks for communicating how you are thinking about this.  It helps me a great deal and I appreciate the effort here. 

Let's make this a draft.  And continue privately if that helps.

Just so you know, there were two things that in the back of my mind I am holding off on right now, but would like to see as part of going over the Form/Box/Formatter reworking. 

The first is to get the for function operators to use MathicsScanner tables and use the unicode Unicode  symbol from those tables when that is appropriate. Likewise for (ams)TeX symbols. This has been something we've both kind of known about for a long time, and I have just been putting it off for when we get around to Form and formatting work which is sometime in this phase (although it doesn't have to be right now and cut be done later).

The second thing that i wanted to do was write [DiscretePlot](https://reference.wolfram.com/language/ref/DiscretePlot.html) which seems pretty simple to do. 

However since that also has an arm in formatting where I expect things to get moved around, I was holding off on that as well.

For me, I think the next non-cleanup thing to do is to redo how `Part` works so that it uses `ListExpression` and works in a more functional way.

Reply:
> 
> Just so you know, there were two things that in the back of my mind I am holding off on right now, but would like to see as part of going over the Form/Box/Formatter reworking.
> 
> The first is to get the for function operators to use MathicsScanner tables and use the Unicode symbols from those tables when that is appropriate. Likewise for (ams)TeX symbols. This has been something we've both kind of known about for a long time, and I have just been putting it off for when we get around to Form and formatting work which is sometime in this phase (although it doesn't have to be right now and can be done later).
> 


> The second thing that i wanted to do was write [DiscretePlot](https://reference.wolfram.com/language/ref/DiscretePlot.html) which seems pretty simple to do.
> 

How is this related to the formatting refactor?

> However since that also has an arm in formatting where I expect things to get moved around, I was holding off on that as well.
> 
> For me, I think the next non-cleanup thing to do is to redo how `Part` works so that it uses `ListExpression` and works in a more functional way.

Same question.

Reply:
`Part` has nothing to do with formatting refactor. It was just a heads up. (You have a tendency to work scattered so declaring what I intend to do I am hoping make sure we don't collide)

As for `DiscretePlot` I looked again given where this is going I see that it isn't related at all. I don't remember the last time I looked (before release) why I thought it was likely to be an areas where there was going to be a lot of motion.

Reply:
> Below, `_BoxedString()` doesn't belong here. In fact it is the only class where everything else is a function
> 
> The boxing class belongs in the box module like all of the others boxing routines. And I don't think the box module should be under `builtin`.
> 
> Here is a suggested reorganization. Under core, we have a `form` (both `convert` and `form` are intended to be imperative verbs not nouns as we did in summary text, as in: you (we, they, or I) convert, you (or we, they, or I) `form`.
> 
> `parser` is a noun, but if things were done from scratch I would have suggested `parse` instead of `parser`. If you think it worth doing, then please do.
> 
> Under form we have "box" (also a verb), and low-level or final format (as opposed to the top level form). Right now this lower-level `format` is not under anything but if we create a `form` i I think it should be moved under that.
> 
> The core module then has parse, evaluate (not currently a module though) and form modules. The form module has `box` and `format` modules.
> 
> If you wanted to write a backend to handle some new `Form` you would model it on the this new form module.
> 
> What do you think?

Regarding `_BoxedString`, this is something that should go away before the next release. I put it in order to start the format decoupling without doing a very drastic change in the interface, but each time I look at it, I feel more regret.

Reply:
Great! Something like this is what I was hoping to hear. Yes, I understand we can't do everything all at once.